### DISCUSSION TITLE
Feature Request: Native dynamic data loaders/generators

### DISCUSSION THREAD

**Discussion by mayureshagashe2105** [google/flax#2315] [Ideas]
To my knowledge, there is no native support for this feature. Can the data loaders/ data generators be added which will dynamically read in the data at the training time using python's native generators?
These can have functionalities like:

- read from disk
- read from dataframe
- patch based learning
- etc.

Similar to [keras.utils.Sequence](https://github.com/keras-team/keras/blob/v2.9.0/keras/utils/data_utils.py#L425-L501) or [torch.utils.data.DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader)

If yes, I would like to contribute.
Thank you.

**Comment by cgarciae** [comment]
Hey! This is not a official stance but I think that the JAX ecosystem has opted to use either [tf.data](https://www.tensorflow.org/guide/data), Pytorch DataLoaders, or any pipeline really as long as you can efficiently convert it into numpy and pass it to JAX. With `tf.data` is pretty easy using [as_numpy_iterator](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#as_numpy_iterator), its also better tested as its used within Google. For Pytorch DataLoaders you can just convert to numpy first before feeding it into JAX.

**Comment by jlperla** [comment]
Curious if this unofficial stance has changed?  I found https://github.com/birkhoffg/jax-dataloader to have a nice, simple interface (especially for pure JAX data).  The alternatives of tf, pythorch, etc. are pretty massive dependencies and many introductory users of ML find the custom dataloader code a little scary.

Would a clean, simple JAX-only dataloader which uses the `rngs=rngs` arguments common to NNX be an acceptable small utility to add?
